### PR TITLE
parse arbitrary datetime strings

### LINKED ISSUE #150: strftime formatting
Thanks for all the hard work on this package!

I'd like to make a request, similar to #62 but maybe a bit simpler. I'd like to be able to specify how I would like a datetime to be parsed. For example:

```python
from parse import parse
parse("my date:{my_date:%Y%M%d}", "my date:20200403").groups = {"my_date": datetime.datetime(2020, 4, 3)}
```

You'd need to construct a regex expression from a date format:

```python
dt_format_to_regex = {symbol: "[0-9]{2}" for symbol in "ymdIMSUW"}
dt_format_to_regex.update({"-" + symbol: "[0-9]{1,2}" for symbol in "ymdIMS"})

dt_format_to_regex.update(
    {
        "a": "[Sun|Mon|Tue|Wed|Thu|Fri|Sat]",
        "A": "[Sunday|Monday|Tuesday|Wednesday|Thursday|Friday|Saturday]",
        "Y": "[0-9]{4}",
        "H": "[0-9]{1,2}",
        "B": "[January|February|March|April|May|June|July|August|September|October|November|December]",
        "b": "[Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec]",
        "f": "[0-9]{6}",
        "p": "[AM|PM]",
        "z": "[+|-][0-9]{4}",
        "j": "[0-9]{3}",
        "-j": "[0-9]{1,3}",
    }
)


def get_regex_for_datetime_format(format_):
    regex = copy(format_)
    for k, v in dt_format_to_regex.items():
        regex = regex.replace(f"%{k}", v)
    return regex
```

and you'd need to parse the string into a datetime, which is actually quite easy because you can just rely on `datetime.datetime.strptime`:

```python
self._type_conversions[group] = lambda x: datetime.strptime(x, format)
```

I guess the remaining bit would be to determine if the formatting is of a datetime, but that doesn't seem that hard to me. As a first pass, you could just look for "%Y" in the format string. Technically `datetime.datetime.strptime` does allow you to parse datestrings that do not have a year, but that results are non-sensical, so it would be fine to not support them:

```python
print(datetime.datetime.strptime("Sun", "%a"))
print(datetime.datetime.strptime("Mon", "%a"))
```
```
datetime.datetime(1900, 1, 1, 0, 0)
datetime.datetime(1900, 1, 1, 0, 0)
```

### PR DESCRIPTION
**Pull request by bendichter**
fix #150

### CODE REVIEW

**Comment by bendichter** [comment]
@r1chardj0n3s , I would love to hear your thoughts on this. It would be very useful for my application and I'd like to avoid splintering if I can

**Comment by bendichter** [comment]
ping @r1chardj0n3s

**Comment by bendichter** [comment]
ping @r1chardj0n3s

**Comment by wimglenn** [comment]
What's up with renaming `parse.py` to `fparse.py`? That's a huge backwards break and I don't understand the reason, there's no explanation in the commit messages

**Comment by bendichter** [comment]
@wimglenn this was a while ago but I think that was a mistake. @r1chardj0n3s never responded so I made a package of my own [here](https://github.com/catalystneuro/fparse) and probably just copy/pasted to aggressively. If there is serious consideration of this feature I'd be happy to fix it and the merge conflicts.

File: README.rst [hunk 1]
```diff
@@ -5,7 +5,7 @@ Parse strings using a specification based on the Python format() syntax.
 The module is set up to only export ``parse()``, ``search()``, ``findall()``,
 and ``with_pattern()`` when ``import \*`` is used:
 
->>> from parse import *
+>>> from fparse import *
```
Comment (bendichter) [inline README.rst:8]:
```suggestion
>>> from parse import *
```

File: README.rst [hunk 2]
```diff
@@ -35,7 +35,7 @@ compile it once:
 
 .. code-block:: pycon
 
-    >>> from parse import compile
+    >>> from fparse import compile
```
Comment (bendichter) [inline README.rst:38]:
```suggestion
    >>> from parse import compile
```

File: fparse.py [hunk 3]
```diff
@@ -5,7 +5,7 @@
 The module is set up to only export ``parse()``, ``search()``, ``findall()``,
 and ``with_pattern()`` when ``import \*`` is used:
 
->>> from parse import *
+>>> from fparse import *
```
Comment (bendichter) [inline fparse.py:8]:
```suggestion
>>> from parse import *
```

File: fparse.py [hunk 4]
```diff
@@ -35,7 +35,7 @@
 
 .. code-block:: pycon
 
-    >>> from parse import compile
+    >>> from fparse import compile
```
Comment (bendichter) [inline fparse.py:38]:
```suggestion
    >>> from parse import compile
```

File: setup.py [hunk 5]
```diff
@@ -4,14 +4,14 @@
 
 from setuptools import setup
 
-from parse import __version__, __doc__
+from fparse import __version__, __doc__
```
Comment (bendichter) [inline setup.py:7]:
```suggestion
from parse import __version__, __doc__
```

File: setup.py [hunk 6]
```diff
@@ -4,14 +4,14 @@
...
 
 with open('README.rst', 'w') as f:
     f.write(__doc__)
 
 # perform the setup action
 setup(
-    name = "parse",
+    name = "fparse",
```
Comment (bendichter) [inline setup.py:14]:
```suggestion
    name = "parse",
```

**Comment by bendichter** [comment]
Closed in favor of #165

**Pull request closed without merging**